### PR TITLE
#674 - Added notification model

### PR DESCRIPTION
**Pull request by raduiordache**
What is there:
- Added the Notification model;
- possibility to retrieve the notifications;
- possibility to create a new notification;
- possibility to delete a notification;
- added Activity.share_link() function;
- added Activity.share_pdf() function;
- wrote tests on it
- cassettes;

What is not there:
- share_pdf() with appendices;

### CODE REVIEW

**Comment by raduiordache** [comment]
@JelleBoersma, @jberends  - can you have a look on my changes and see whether I missed something?

**Comment by JelleBoersma** [comment]
- [ ] Add to changelog
- [ ] Fix typing issues

File: pykechain/defaults.py [hunk 1]
```diff
@@ -144,4 +148,8 @@
     'widget': {'fields': ",".join(
         ['id', 'name', 'ref', 'created_at', 'updated_at', 'title', 'widget_type', 'meta', 'order', 'activity_id',
          'parent_id', 'progress', 'has_subwidgets', 'scope_id'])},
+    'notifications': {'fields': ",".join(
+        ['id', 'subject', 'status', 'message', 'team', 'created_at', 'updated_at'])},
```
Comment (jberends) [inline pykechain/defaults.py]: options here as well to be sure

File: pykechain/client.py [hunk 2]
```diff
@@ -2847,3 +2848,78 @@ def update_properties(self, properties: List[Dict]) -> List['AnyProperty']:
         properties = [Property2.create(client=self, json=js) for js in response.json()['results']]
 
         return properties
+
+
+    def notifications(self, pk=None, **kwargs):
+        """
+
+        :param pk:
```
Comment (jberends) [inline pykechain/client.py]: fill in documentation

File: pykechain/client.py [hunk 3]
```diff
@@ -2847,3 +2848,78 @@ def update_properties(self, properties: List[Dict]) -> List['AnyProperty']:
...
+        response = self._request('GET', self._build_url('notifications'), params=request_params)
+
+        data = response.json()
+
+        return [Notification(notification, client=self) for notification in data['results']]
+
+    def notification(self, pk=None, **kwargs):
+        """
+
+        :param pk:
```
Comment (jberends) [inline pykechain/client.py]: docs

File: pykechain/client.py [hunk 4]
```diff
@@ -2847,3 +2848,78 @@ def update_properties(self, properties: List[Dict]) -> List['AnyProperty']:
...
+
+        if len(_notifications) == 0:
+            raise NotFoundError("No notification found with id:{}".format(criteria))
+        if len(_notifications) != 1:
+            raise MultipleFoundError("Multiple users fit criteria:{}".format(criteria))
+
+        return _notifications[0]
+
+    def create_notification(self, **kwargs):
+        query_params = kwargs
```
Comment (jberends) [inline pykechain/client.py]: docs

File: pykechain/enums.py [hunk 5]
```diff
@@ -909,3 +909,40 @@ class WidgetTitleValue(Enum):
     DEFAULT = 'Default'
     NO_TITLE = 'No title'
     CUSTOM_TITLE = 'Custom title'
+
+
+class NotificationStatus(Enum):
+    """
+    Options to retrieve a Notification based on its status.
+
```
Comment (jberends) [inline pykechain/enums.py]: maybe show the status flow here. I guess there is somewhere an ascii drawing on the flow let met get it

```transition_matrix = {
            NotificationStatus.DRAFT: [NotificationStatus.READY],
            NotificationStatus.READY: [NotificationStatus.PROCESSING],
            NotificationStatus.PROCESSING: [NotificationStatus.COMPLETED, NotificationStatus.FAILED],
            NotificationStatus.COMPLETED: [NotificationStatus.ARCHIVED],
            NotificationStatus.FAILED: [NotificationStatus.ARCHIVED],
            NotificationStatus.ARCHIVED: []
        }
```

```
    normal lifecycle:

    - DRAFT, when a message is first saved to the backend and the status is still in draft.
        next states: READY
    - READY: when the message is ready for processing, it is complete and is to be processed
        next states: PROCESSING
    - PROCESSING: when the message is in the process of being send out
        next states: COMPLETED, FAILED
    - COMPLETED: when the message is successfully sent out
        next states: ARCHIVED
    - FAILED: when the message is not successfully sent out
        next states: ARCHIVED
    - ARCHIVED: when the message is archives and waiting for its deletion against a certain retention policy
        next states: None
```

File: tests/test_activities.py [hunk 6]
```diff
@@ -658,3 +658,35 @@ def test_activity2_download_as_pdf_async(self):
...
             self.assertTrue(pdf_file_called_after_activity)
+
+    @skip
+    def test_activity2_share_link(self):
+        activity_name = 'Task - Form + Tables + Service'
+        message = 'EXAMPLE_MESSAGE'
+        subject = 'EXAMPLE_SUBJECT'
+        recipient_users = [96]
+
+        activity = self.project.activity(name=activity_name)
```
Comment (jberends) [inline tests/test_activities.py]: send it to no-reply+pykechaintesting@ke-chain.com

File: tests/test_activities.py [hunk 7]
```diff
@@ -658,3 +658,35 @@ def test_activity2_download_as_pdf_async(self):
...
+                            recipient_users=recipient_users)
+
+    @skip
+    def test_activity2_share_pdf(self):
+        activity_name = 'Task - Form + Tables + Service'
+        message = 'EXAMPLE_MESSAGE'
+        subject = 'EXAMPLE_SUBJECT'
+        paper_size = PaperSize.A2
+        paper_orientation = PaperOrientation.PORTRAIT
+        recipient_users = [96]
```
Comment (jberends) [inline tests/test_activities.py]: too hard coded! because now we cannot run this locally

File: pykechain/defaults.py (hunk 1 above)
Comment (raduiordache) [reply pykechain/defaults.py]: Done

File: pykechain/client.py (hunk 2 above)
Comment (raduiordache) [reply pykechain/client.py]: Done

File: pykechain/client.py (hunk 3 above)
Comment (raduiordache) [reply pykechain/client.py]: Done

File: pykechain/client.py (hunk 4 above)
Comment (raduiordache) [reply pykechain/client.py]: done

File: pykechain/enums.py (hunk 5 above)
Comment (raduiordache) [reply pykechain/enums.py]: Gedaan

File: tests/test_activities.py (hunk 6 above)
Comment (raduiordache) [reply tests/test_activities.py]: Done

**Comment by raduiordache** [comment]
@jberends - please re-review, i have implemented all the changes you and @JelleBoersma  asked for

File: pykechain/models/activity2.py [hunk 8]
```diff
@@ -920,3 +921,110 @@ def move(self, parent, classification=None):
         :raises APIError: if an Error occurs.
         """
         return self._client.move_activity(self, parent, classification=classification)
+
+    def share_link(self, subject=None, message=None, recipient_users=None):
```
Comment (JelleBoersma) [inline pykechain/models/activity2.py]: If some inputs are required, why do they have default=None values?

File: pykechain/models/activity2.py [hunk 9]
```diff
@@ -920,3 +921,110 @@ def move(self, parent, classification=None):
...
+        )
+
+        url = self._client._build_url('notification_share_activity_link')
+
+        response = self._client._request('POST', url, data=params)
+
+        if response.status_code != requests.codes.created:  # pragma: no cover
+            raise APIError("Could not share the link to Activity, {}:\n\n{}'".format(str(response), response.json()))
+
+    def share_pdf(self, subject=None, message=None, recipient_users=None, paper_size=PaperSize.A4,
```
Comment (JelleBoersma) [inline pykechain/models/activity2.py]: Same as earlier comment: if a value must be provided a default value is not advised

File: pykechain/models/activity2.py (hunk 8 above)
Comment (jberends) [reply pykechain/models/activity2.py]: good point

```suggestion
    def share_link(self, subject, message, recipient_users):
```

File: pykechain/models/activity2.py [hunk 10]
```diff
@@ -920,3 +921,110 @@ def move(self, parent, classification=None):
...
+        :param message: message of email
+        :type message: basestring
+        :param recipient_users: users that will receive the email
+        :type recipient_users: list(Union(User, Id))
+        :raises IllegalArgumentError: if no 'message' is specified
+        :raises IllegalArgumentError: if no 'subject' is specified
+        :raises IllegalArgumentError: if no 'recipient_users' is specified
+        :raises APIError: if an Error occurs.
+        """
+        if not message:
```
Comment (jberends) [inline pykechain/models/activity2.py]: remove as it is required input

File: pykechain/models/activity2.py (hunk 9 above)
Comment (jberends) [reply pykechain/models/activity2.py]:
```suggestion
    def share_pdf(self, subject, message, recipient_users, paper_size=PaperSize.A3,
```

File: tests/test_activities.py [hunk 11]
```diff
@@ -658,3 +658,35 @@ def test_activity2_download_as_pdf_async(self):
...
+
+        activity = self.project.activity(name=activity_name)
+
+        activity.share_pdf(subject=subject,
+                           message=message,
+                           recipient_users=recipient_users,
+                           paper_size=paper_size,
+                           paper_orientation=paper_orientation,
+                           include_appendices=False
+                           )
```
[the remark this answers is not in the document]
Comment (jberends) [reply tests/test_activities.py]: still to do this one

- [ ] check response status code

**Pull request merged**